### ISSUE TITLE
build fails on dune locking? due to sandboxing

### ISSUE THREAD

**Issue by Blaisorblade** [coq/platform#189]
On current `main` (8568054b1926352d71bc81f831351c01855ed3ec), the build fails whether the dune daemon is stopped or started — the daemon fails to start, and if it exists we fail to connect to it. Opam sandboxing might be involved.

```
$ git describe
2021.09.0-54-g8568054
$ dune cache stop
$ time ./coq_platform_make.sh -extent=b -pick=8.14~2021.11 -parallel=p -jobs=16
===== CHECKING VERSION OF INSTALLED OPAM =====
Found opam 2.1.1 - good!
/opt/homebrew/bin/opam
===== opam already initialized =====
===== CREATE OPAM SWITCH =====
[__coq-platform.2021.11.0.patch_ocaml] Initialised
[__coq-platform.2021.11.0.patch_coq-released] Initialised
[__coq-platform.2021.11.0.patch_coq-dev] Initialised
[coq-released] no changes from https://coq.inria.fr/opam/released
[coq-core-dev] no changes from https://coq.inria.fr/opam/core-dev
[coq-extra-dev] no changes from https://coq.inria.fr/opam/extra-dev

<><> Installing new switch packages <><><><><><><><><><><><><><><><><><><><>  🐫
Switch invariant: ["ocaml-base-compiler" {= "4.10.2"}]

<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><>  🐫
∗ installed base-bigarray.base
∗ installed base-threads.base
∗ installed base-unix.base
⬇ retrieved ocaml-base-compiler.4.10.2  (cached)
∗ installed ocaml-base-compiler.4.10.2
∗ installed ocaml-config.1
∗ installed ocaml.4.10.2
Done.
=== OPAM REPOSITORIES ===
[NOTE] These are the repositories in use by the current switch. Use '--all' to see all configured repositories.

<><> Repository configuration for switch __coq-platform.2021.11.0~8.14~2021.11
 1 __coq-platform.2021.11.0.patch_coq-released file:///Users/pgiarrusso1/git-bedrock/coq-platform/opam/opam-coq-archive/released
 2 __coq-platform.2021.11.0.patch_ocaml        file:///Users/pgiarrusso1/git-bedrock/coq-platform/opam/opam-repository
 3 coq-released                                https://coq.inria.fr/opam/released
 4 default                                     https://opam.ocaml.org
=== OPAM PACKAGES ===
# Packages matching: installed
# Name              # Installed # Synopsis
base-bigarray       base
base-threads        base
base-unix           base
ocaml               4.10.2      The OCaml compiler (virtual package)
ocaml-base-compiler 4.10.2      Official release 4.10.2
ocaml-config        1           OCaml Switch Configuration
Cleaning up switch __coq-platform.2021.11.0~8.14~2021.11
===== UPDATE OPAM REPOSITORIES =====

<><> Updating package repositories ><><><><><><><><><><><><><><><><><><><><>  🐫
[__coq-platform.2021.11.0.patch_ocaml] no changes from file:///Users/pgiarrusso1/git-bedrock/coq-platform/opam/opam-repository
[__coq-platform.2021.11.0.patch_coq-released] no changes from file:///Users/pgiarrusso1/git-bedrock/coq-platform/opam/opam-coq-archive/released
[coq-released] no changes from https://coq.inria.fr/opam/released
[default] synchronised from https://opam.ocaml.org
Now run 'opam upgrade' to apply any package updates.
POL="$POL"'(allow network* (remote unix))'
===== INSTALL PREREQUISITES =====
[WARNING] set was deprecated in version 2.1 of the opam CLI. Use opam var instead or set OPAMCLI environment variable to 2.0.
Added 'jobs: "16"' to field variables in switch __coq-platform.2021.11.0~8.14~2021.11
===== INSTALL OPAM PACKAGES (PARALLEL) =====
The following actions will be performed:
  ∗ install ocamlfind      1.9.1  [required by coq]
  ∗ install dune           2.9.1  [required by coq]
  ∗ install conf-gmp       3      [required by zarith]
  ∗ install conf-findutils 1      [required by coq]
  ∗ install zarith         1.12   [required by coq]
  ∗ install coq            8.14.0
===== ∗ 6 =====

<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><>  🐫
⬇ retrieved ocamlfind.1.9.1  (cached)
⬇ retrieved zarith.1.12  (cached)
∗ installed conf-gmp.3
∗ installed conf-findutils.1
⬇ retrieved dune.2.9.1  (cached)
⬇ retrieved coq.8.14.0  (https://opam.ocaml.org/cache)
∗ installed ocamlfind.1.9.1
∗ installed zarith.1.12
∗ installed dune.2.9.1
[ERROR] The compilation of coq.8.14.0 failed at "./configure -configdir /Users/pgiarrusso1/.opam/__coq-platform.2021.11.0~8.14~2021.11/lib/coq/config -prefix /Users/pgiarrusso1/.opam/__coq-platform.2021.11.0~8.14~2021.11 -mandir
        /Users/pgiarrusso1/.opam/__coq-platform.2021.11.0~8.14~2021.11/man -docdir /Users/pgiarrusso1/.opam/__coq-platform.2021.11.0~8.14~2021.11/doc/coq -libdir /Users/pgiarrusso1/.opam/__coq-platform.2021.11.0~8.14~2021.11/lib/coq -datadir
        /Users/pgiarrusso1/.opam/__coq-platform.2021.11.0~8.14~2021.11/share/coq -coqide no -native-compiler no".

#=== ERROR while compiling coq.8.14.0 =========================================#
# context              2.1.1 | macos/arm64 | ocaml-base-compiler.4.10.2 | https://opam.ocaml.org#28e817d4
# path                 ~/.opam/__coq-platform.2021.11.0~8.14~2021.11/.opam-switch/build/coq.8.14.0
# command              ~/.opam/opam-init/hooks/sandbox.sh build ./configure -configdir /Users/pgiarrusso1/.opam/__coq-platform.2021.11.0~8.14~2021.11/lib/coq/config -prefix /Users/pgiarrusso1/.opam/__coq-platform.2021.11.0~8.14~2021.11 -mandir /Users/pgiarrusso1/.opam/__coq-platform.2021.11.0~8.14~2021.11/man -docdir /Users/pgiarrusso1/.opam/__coq-platform.2021.11.0~8.14~2021.11/doc/coq -libdir /Users/pgiarrusso1/.opam/__coq-platform.2021.11.0~8.14~2021.11/lib/coq -datadir /Users/pgiarrusso1/.opam/__coq-platform.2021.11.0~8.14~2021.11/share/coq -coqide no -native-compiler no
# exit-code            1
# env-file             ~/.opam/log/coq-41428-d0e8d3.env
# output-file          ~/.opam/log/coq-41428-d0e8d3.out
### output ###
# Error: too many retries (100): waiting for beacon file
# "/Users/pgiarrusso1/.cache/dune/db/runtime/dune-cache-daemon/port" to be
# locked
# Error: empty output starting cache



<><> Error report <><><><><><><><><><><><><><><><><><><><><><><><><><><><><>  🐫
┌─ The following actions failed
│ λ build coq 8.14.0
└─
┌─ The following changes have been performed
│ ∗ install conf-findutils 1
│ ∗ install conf-gmp       3
│ ∗ install dune           2.9.1
│ ∗ install ocamlfind      1.9.1
│ ∗ install zarith         1.12
└─

The former state can be restored with:
    /opt/homebrew/bin/opam switch import "/Users/pgiarrusso1/.opam/__coq-platform.2021.11.0~8.14~2021.11/.opam-switch/backup/state-20211202203948.export"

real	2m47.329s
user	4m31.335s
sys	1m25.210s
```

```
$ dune cache stop
$ time ./coq_platform_make.sh -extent=b -pick=8.14~2021.11 -parallel=p -jobs=16
127.0.0.1:62972
$ time ./coq_platform_make.sh -extent=b -pick=8.14~2021.11 -parallel=p -jobs=16
================================ OPAM SWITCH =================================
The Coq Platform creates the opam switch __coq-platform.2021.11.0~8.14~2021.11.

Apparently this switch already exists. It is recommended to delete the switch,
so that you get a clean and well defined result.

For incremental builds after a failure, e.g. cause of RAM size issues, it is
no problem to keep the switch.
================================ OPAM SWITCH =================================
Shall the existing switch be kept (k) or deleted (d) ? (k/d/c=cancel) k

===== CHECKING VERSION OF INSTALLED OPAM =====
Found opam 2.1.1 - good!
/opt/homebrew/bin/opam
===== opam already initialized =====
===== opam switch already exists =====
=== OPAM REPOSITORIES ===
[NOTE] These are the repositories in use by the current switch. Use '--all' to see all configured repositories.

<><> Repository configuration for switch __coq-platform.2021.11.0~8.14~2021.11
 1 __coq-platform.2021.11.0.patch_coq-released file:///Users/pgiarrusso1/git-bedrock/coq-platform/opam/opam-coq-archive/released
 2 __coq-platform.2021.11.0.patch_ocaml        file:///Users/pgiarrusso1/git-bedrock/coq-platform/opam/opam-repository
 3 coq-released                                https://coq.inria.fr/opam/released
 4 default                                     https://opam.ocaml.org
=== OPAM PACKAGES ===
# Packages matching: installed
# Name              # Installed # Synopsis
base-bigarray       base
base-threads        base
base-unix           base
conf-findutils      1           Virtual package relying on findutils
conf-gmp            3           Virtual package relying on a GMP lib system installation
dune                2.9.1       Fast, portable, and opinionated build system
ocaml               4.10.2      The OCaml compiler (virtual package)
ocaml-base-compiler 4.10.2      Official release 4.10.2
ocaml-config        1           OCaml Switch Configuration
ocamlfind           1.9.1       A library manager for OCaml
zarith              1.12        Implements arithmetic and logical operations over arbitrary-precision integers
Cleaning up switch __coq-platform.2021.11.0~8.14~2021.11
===== UPDATE OPAM REPOSITORIES =====

<><> Updating package repositories ><><><><><><><><><><><><><><><><><><><><>  🐫
[__coq-platform.2021.11.0.patch_coq-released] no changes from file:///Users/pgiarrusso1/git-bedrock/coq-platform/opam/opam-coq-archive/released
[__coq-platform.2021.11.0.patch_ocaml] no changes from file:///Users/pgiarrusso1/git-bedrock/coq-platform/opam/opam-repository
POL="$POL"'(allow network* (remote unix))'
===== INSTALL PREREQUISITES =====
[WARNING] set was deprecated in version 2.1 of the opam CLI. Use opam var instead or set OPAMCLI environment variable to 2.0.
Added 'jobs: "16"' to field variables in switch __coq-platform.2021.11.0~8.14~2021.11
===== INSTALL OPAM PACKAGES (PARALLEL) =====
The following actions will be performed:
  ∗ install coq 8.14.0

<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><>  🐫
⬇ retrieved coq.8.14.0  (cached)
[ERROR] The compilation of coq.8.14.0 failed at "./configure -configdir /Users/pgiarrusso1/.opam/__coq-platform.2021.11.0~8.14~2021.11/lib/coq/config -prefix /Users/pgiarrusso1/.opam/__coq-platform.2021.11.0~8.14~2021.11 -mandir
        /Users/pgiarrusso1/.opam/__coq-platform.2021.11.0~8.14~2021.11/man -docdir /Users/pgiarrusso1/.opam/__coq-platform.2021.11.0~8.14~2021.11/doc/coq -libdir /Users/pgiarrusso1/.opam/__coq-platform.2021.11.0~8.14~2021.11/lib/coq -datadir
        /Users/pgiarrusso1/.opam/__coq-platform.2021.11.0~8.14~2021.11/share/coq -coqide no -native-compiler no".

#=== ERROR while compiling coq.8.14.0 =========================================#
# context              2.1.1 | macos/arm64 | ocaml-base-compiler.4.10.2 | https://opam.ocaml.org#28e817d4
# path                 ~/.opam/__coq-platform.2021.11.0~8.14~2021.11/.opam-switch/build/coq.8.14.0
# command              ~/.opam/opam-init/hooks/sandbox.sh build ./configure -configdir /Users/pgiarrusso1/.opam/__coq-platform.2021.11.0~8.14~2021.11/lib/coq/config -prefix /Users/pgiarrusso1/.opam/__coq-platform.2021.11.0~8.14~2021.11 -mandir /Users/pgiarrusso1/.opam/__coq-platform.2021.11.0~8.14~2021.11/man -docdir /Users/pgiarrusso1/.opam/__coq-platform.2021.11.0~8.14~2021.11/doc/coq -libdir /Users/pgiarrusso1/.opam/__coq-platform.2021.11.0~8.14~2021.11/lib/coq -datadir /Users/pgiarrusso1/.opam/__coq-platform.2021.11.0~8.14~2021.11/share/coq -coqide no -native-compiler no
# exit-code            1
# env-file             ~/.opam/log/coq-49018-32d7f5.env
# output-file          ~/.opam/log/coq-49018-32d7f5.out
### output ###
# Error: connect: : Operation not permitted



<><> Error report <><><><><><><><><><><><><><><><><><><><><><><><><><><><><>  🐫
┌─ The following actions failed
│ λ build coq 8.14.0
└─
╶─ No changes have been performed

real	0m20.735s
user	0m9.202s
sys	0m8.676s
```

**Comment by Blaisorblade** [comment]
In particular, see `POL="$POL"'(allow network* (remote unix))'`? That seems from /Users/pgiarrusso1/.opam/opam-init/hooks/sandbox.sh...

**Comment by Blaisorblade** [comment]
I'm trying 8.13, since that older Coq should not rely on dune during installation — using `-pick=8.13~2021.11`.

**Comment by Blaisorblade** [comment]
I should probably note that I _have_ run dune 3 beta on this profile, but dune 2 still works fine usually.

I'm not blocked on this, but cc @ejgallego @rgrinberg for any insights.

**Comment by rgrinberg** [comment]
> On current main (8568054), the build fails whether the dune daemon is stopped or started — the daemon fails to start, and if it exists we fail to connect to it. Opam sandboxing might be involved.

Not sure what this means, there is no more daemon in the main branch. It was removed in favor of a simpler architecture. 

Try doing an strace to see where the `connect` is coming from.

**Comment by MSoegtropIMC** [comment]
@rgrinberg: we are discussing Coq Platform main, not dune main. Coq Platform currently uses the latest release dune opam package.

**Comment by rgrinberg** [comment]
Thanks for clarifying. In that case, the opam sandbox is definitely the issue.

**Comment by MSoegtropIMC** [comment]
So I guess this issue will be fixed by an update to dune. Are there estimates when this will be available?

**Comment by rgrinberg** [comment]
Should be end of year or very early next year. 

You will still need to tweak opam's sandboxing somehow though. Otherwise, it will still fail when dune will try to write to the shared cache directly.

**Comment by Blaisorblade** [comment]
Does this mean installing Coq 8.14 on its own would fail? I guess one must enable caching first, so there's a workaround.

**Comment by MSoegtropIMC** [comment]
@Blaisorblade : I understood you such that this issue only happens if one does some special advanced dune stuff. Did I get this wrong?

If this is so, I would suggest we say this special advanced dune stuff is not supported until the version where this is finalized is released.

**Comment by Blaisorblade** [comment]
@MSoegtropIMC Yes; I'd say dune caching seems dune's killer feature, but clearly it's uncommon if nobody reported this yet. And this problem is not with the platform but with Coq.

**Comment by Blaisorblade** [comment]
https://coq.zulipchat.com/#narrow/stream/237977-Coq-users/topic/dune.20cache.20breaks.20Coq.208.2E14.28.2E1.29.20.28on.20Mac.3F.29 suggests the issue might be platform-specific; I have to investigate.

**Comment by Blaisorblade** [comment]
Closing as duplicate of https://github.com/coq/coq/issues/15138.
